drivers/interrupt_controller/intc_ite_it8xxx2_v2.c

```diff
@@ -0,0 +1,245 @@
+
+/* Interrupt status register */
+#define IT8XXX2_INTC_ISR(g)	ECREG(IT8XXX2_INTC_BASE_SHIFT(g) + \
+				((g) < 4 ? 0x0 : 0x4))
+/* Interrupt enable register */
+#define IT8XXX2_INTC_IER(g)	ECREG(IT8XXX2_INTC_BASE_SHIFT(g) + \
+				((g) < 4 ? 0x1 : 0x5))
+/* Interrupt edge/level triggered mode register */
+#define IT8XXX2_INTC_IELMR(g)	ECREG(IT8XXX2_INTC_BASE_SHIFT(g) + \
+				((g) < 4 ? 0x2 : 0x6))
+/* Interrupt polarity register */
+#define IT8XXX2_INTC_IPOLR(g)	ECREG(IT8XXX2_INTC_BASE_SHIFT(g) + \
+				((g) < 4 ? 0x3 : 0x7))
+
+#define IT8XXX2_INTC_GROUP_CNT      24
+#define MAX_REGISR_IRQ_NUM          8
+#define IVECT_OFFSET_WITH_IRQ       0x10
+
+/* Interrupt number of INTC module */
+static uint8_t intc_irq;
+static uint8_t ier_setting[IT8XXX2_INTC_GROUP_CNT];
+
+void ite_intc_save_and_disable_interrupts(void)
+{
+	volatile uint8_t _ier __unused;
+
+	/* Disable global interrupt for critical section */
+	unsigned int key = irq_lock();
+
+	/* Save and disable interrupts */
+	for (int i = 0; i < IT8XXX2_INTC_GROUP_CNT; i++) {
+		ier_setting[i] = IT8XXX2_INTC_IER(i);
+		IT8XXX2_INTC_IER(i) = 0;
+	}
+
+	/*
+	 * This load operation will guarantee the above modification of
+	 * SOC's register can be seen by any following instructions.
+	 * Note: Barrier instruction can not synchronize chip register,
+	 * so we introduce workaround here.
+	 */
+	_ier = IT8XXX2_INTC_IER(IT8XXX2_INTC_GROUP_CNT - 1);
+
+	irq_unlock(key);
+}
+
+void ite_intc_restore_interrupts(void)
+{
+	/*
+	 * Ensure the highest priority interrupt will be the first fired
+	 * interrupt when soc is ready to go.
+	 */
+	unsigned int key = irq_lock();
+
+	/* Restore interrupt state */
+	for (int i = 0; i < IT8XXX2_INTC_GROUP_CNT; i++) {
+		IT8XXX2_INTC_IER(i) = ier_setting[i];
+	}
+
+	irq_unlock(key);
+}
+
+void ite_intc_isr_clear(unsigned int irq)
+{
+	uint32_t group, index;
+
+	if (irq > CONFIG_NUM_IRQS) {
+		return;
+	}
+
+	group = irq / MAX_REGISR_IRQ_NUM;
+	index = irq % MAX_REGISR_IRQ_NUM;
+
+	IT8XXX2_INTC_ISR(group) = BIT(index);
+}
+
+void __soc_ram_code ite_intc_irq_enable(unsigned int irq)
+{
+	uint32_t group, index;
+
+	if (irq > CONFIG_NUM_IRQS) {
+		return;
+	}
+
+	group = irq / MAX_REGISR_IRQ_NUM;
+	index = irq % MAX_REGISR_IRQ_NUM;
+
+	/* Critical section due to run a bit-wise OR operation */
+	unsigned int key = irq_lock();
+
+	IT8XXX2_INTC_IER(group) |= BIT(index);
+
+	irq_unlock(key);
+}
+
+void __soc_ram_code ite_intc_irq_disable(unsigned int irq)
+{
+	uint32_t group, index;
+	volatile uint8_t _ier __unused;
+
+	if (irq > CONFIG_NUM_IRQS) {
+		return;
+	}
+
+	group = irq / MAX_REGISR_IRQ_NUM;
+	index = irq % MAX_REGISR_IRQ_NUM;
+
+	/* Critical section due to run a bit-wise NAND operation */
+	unsigned int key = irq_lock();
+
+	IT8XXX2_INTC_IER(group) &= ~BIT(index);
+	/*
+	 * This load operation will guarantee the above modification of
+	 * SOC's register can be seen by any following instructions.
+	 */
+	_ier = IT8XXX2_INTC_IER(group);
```

Comment:
You should be able to remove `_ier` and keep only the read, since that's a volatile read. `_ier` forces the unnecessary stack allocation of a variable that gets written but is otherwise unused.

    /*
     * This load operation will guarantee the above modification of
	 * SOC's register can be seen by any following instructions.
	 */
    IT8XXX2_INTC_IER(group);

(Same above in `ite_intc_save_and_disable_interrupts`.)

Reply:
Removed.